trestle/core/profile_resolver.py

```diff
@@ -0,0 +1,650 @@
+            if self._catalog.back_matter is not None and self._catalog.back_matter.resources is not None:
+                for resource in self._catalog.back_matter.resources:
+                    if resource.uuid in needed_uuid_refs:
+                        new_resources.append(resource)
+
+            new_groups: Optional[List[cat.Group]] = list(group_dict.values())
+
+            # should avoid empty lists so set to None if empty
+            new_resources = new_resources if new_resources else None
+            new_groups = new_groups if new_groups else None
+
+            new_cat = cat.Catalog(
+                uuid=str(uuid4()),
+                metadata=self._catalog.metadata,
+                back_matter=common.BackMatter(resources=new_resources),
+                groups=new_groups
+            )
+
+            return new_cat
+
+        def process(self, catalog_iter: Iterable[cat.Catalog]) -> cat.Catalog:
+            """Prune the catalog based on the include rule in the import_."""
+            # this only processes the one catalog yielded by the one import in this pipeline.
+            # it must yield in order to have the merge filter loop over available imported catalogs.
+            self._set_catalog(next(catalog_iter))
+            logger.debug(f'prune yielding catalog {self._catalog.metadata.title} with import {self._import.href}')
+            yield self._prune_catalog()
+
+    class Merge(Pipeline.Filter):
+        """Merge the incoming catalogs according to rules in the profile."""
+
+        def __init__(self, profile: prof.Profile) -> None:
+            """Initialize the class with the profile."""
+            logger.debug('merge filter initialize')
+            self._profile = profile
+
+        def _merge_catalog(self, merged: cat.Catalog, catalog: cat.Catalog) -> cat.Catalog:
+            if merged is None:
+                return catalog
+            if catalog.groups is not None:
+                for group in catalog.groups:
+                    if group.id not in [g.id for g in merged.groups]:
+                        merged.groups.append(cat.Group(id=group.id, title=group.title, controls=[]))
+                    index = [g.id for g in merged.groups].index(group.id)
+                    merged.groups[index].controls.extend(group.controls)
+            return merged
+
+        def process(self, pipelines: List[Pipeline]) -> cat.Catalog:
+            """Merge the incoming catalogs."""
+            # this pulls from import and iterates over the incoming catalogs
+            merged: Optional[cat.Catalog] = None
+            logger.debug(f'merge entering process with {len(pipelines)} pipelines')
+            for pipeline in pipelines:
+                catalog = next(pipeline.process(None))
+                merged = self._merge_catalog(merged, catalog)
+            yield merged
+
+    class Modify(Pipeline.Filter):
+        """Modify the controls based on the profile."""
```

Comment:
The concern that I have for this class is it combines two operations:
- Parameter substitution in the prose
- Modify operations
Do we think this should be done together or separately?

Reply:
From NIST docs:  "The modification phase is an opportunity for a profile to amend, alter or edit controls, and to set parameter values or constrain or qualify parameters as represented in the target. "  I explicitly follow the recommended stages of Import, Merge, Modify but I include a separate filter for Prune.

Reply:
I think the comment I have here is 'Setting' of parameter values *WITHIN* the prose versus outside of the prose is a different operation. I agree setting of the parameter value in the object sense must be done here, however, I was not sure as to whether we wanted to flatten the prose here as well.